src/registration/metric/demons4D.h

```diff
@@ -29,26 +30,40 @@ namespace MR
     namespace Metric
     {
 
-      template <class Im1ImageType, class Im2ImageType, class Im1MaskType, class Im2MaskType> 
+      template <class Im1ImageType, class Im2ImageType, class Im1MaskType, class Im2MaskType>
       class Demons4D { MEMALIGN(Demons4D<Im1ImageType,Im2ImageType,Im1MaskType,Im2MaskType>)
         public:
           Demons4D (default_type& global_energy, size_t& global_voxel_count,
-                     const Im1ImageType& im1_image, const Im2ImageType& im2_image, const Im1MaskType im1_mask, const Im2MaskType im2_mask) :
+                     const Im1ImageType& im1_image, const Im2ImageType& im2_image, const Im1MaskType im1_mask, const Im2MaskType im2_mask,
+                     const vector<MultiContrastSetting>* contrast_settings = nullptr) :
                        global_cost (global_energy),
                        global_voxel_count (global_voxel_count),
                        thread_cost (0.0),
                        thread_voxel_count (0),
                        mutex (new std::mutex),
                        normaliser (0.0),
-                       robustness_parameter (-1.e12),
+                       robustness_parameter (1.e-12), // MP: used to be -1.e12
```

Comment:
Does this require greater investigation / tracking? I.e. was this a bug that requires reporting?

Reply:
I guess Dave put this parameter in the code to prevent floating-point rounding errors in the cost function value to affect the convergence check. I am not sure if it makes a difference in practice and if so I'd expect the code to terminate earlier at roughly the same solution (relevant [code](https://github.com/MRtrix3/mrtrix3/blob/b9c0114afbed727655cb59b80458fdb0987debf8/src/registration/metric/demons.h#L106-L123)).